lib/Plack/Util.pm

```diff
@@ -171,29 +174,57 @@ sub header_iter {
 sub header_get {
     my($headers, $key) = (shift, lc shift);
 
-    my @val;
-    header_iter $headers, sub {
-        push @val, $_[1] if lc $_[0] eq $key;
-    };
+    return () if not @$headers;
+
+    my $i = 0;
+
+    if (wantarray) {
+        $i = -2;
+        return map {
+            $i += 2;
+            $key eq lc $headers->[$i] ? $headers->[$i+1] : ();
+        } 1 .. @$headers/2;
+    }
 
-    return wantarray ? @val : $val[0];
+    while ($i < @$headers) {
+        return $headers->[$i+1] if $key eq lc $headers->[$i];
+        $i += 2;
+    }
+
+    ();
 }
 
 sub header_set {
     my($headers, $key, $val) = @_;
 
-    my($set, @new_headers);
-    header_iter $headers, sub {
-        if (lc $key eq lc $_[0]) {
-            return if $set;
-            $_[1] = $val;
-            $set++;
-        }
-        push @new_headers, $_[0], $_[1];
-    };
+    @$headers = ($key, $val), return if not @$headers;
+
+    my $_key = lc $key;
+    my $i = -2;
 
-    push @new_headers, $key, $val unless $set;
-    @$headers = @new_headers;
+    while (($i += 2) < @$headers) {
+        next if $_key ne lc $headers->[$i];
+
+        $headers->[$i+1] = $val;
+        last;
+    }
+
+    if ($i < @$headers) {
```

Comment:
I think the readability suffers a lot at this point. I understand that you can gain some (or a lot of) performance with this, but hesitate to maintain this bit of code.

Is it possible to make it look less scary?